crates/base/src/rt_worker/worker_ctx.rs

```diff
@@ -83,43 +89,98 @@ async fn handle_request(
     // create a unix socket pair
     let (sender_stream, recv_stream) = UnixStream::pair()?;
     let WorkerRequestMsg {
-        req,
+        mut req,
         res_tx,
         conn_watch,
     } = msg;
 
     let _ = unix_stream_tx.send((recv_stream, conn_watch.clone()));
+    let req_upgrade_type = get_upgrade_type(req.headers());
+    let req_upgrade = req_upgrade_type
+        .clone()
+        .and_then(|it| Some(it).zip(req.extensions_mut().remove::<OnUpgrade>()));
 
     // send the HTTP request to the worker over Unix stream
-    let (mut request_sender, connection) = hyper::client::conn::handshake(sender_stream).await?;
+    let (mut request_sender, connection) = http1::Builder::new()
+        .writev(true)
+        .handshake(sender_stream)
+        .await?;
+
+    let (upgrade_tx, upgrade_rx) = oneshot::channel();
 
     // spawn a task to poll the connection and drive the HTTP state
-    tokio::task::spawn(async move {
-        match connection.without_shutdown().await {
-            Err(e) => {
-                error!("Error in worker connection: {}", e.message(),);
-            }
+    tokio::task::spawn({
+        async move {
+            match connection.without_shutdown().await {
+                Err(e) => {
+                    error!("Error in worker connection: {}", e.message());
+                }
+
+                Ok(parts) => {
+                    if let Some((requested, req_upgrade)) = req_upgrade {
+                        if let Ok((Some(accepted), status)) = upgrade_rx.await {
+                            if status == StatusCode::SWITCHING_PROTOCOLS && accepted == requested {
+                                tokio::spawn(relay_upgraded_request_and_response(
+                                    req_upgrade,
+                                    parts,
+                                ));
 
-            Ok(parts) => {
-                if let Some(mut watcher) = conn_watch {
-                    if watcher.wait_for(|it| *it == ConnSync::Recv).await.is_err() {
-                        error!("cannot track outbound connection correctly");
+                                return;
+                            }
+                        };
+                    }
+
+                    if let Some(mut watcher) = conn_watch {
+                        if watcher.wait_for(|it| *it == ConnSync::Recv).await.is_err() {
+                            error!("cannot track outbound connection correctly");
+                        }
                     }
-                }
 
-                drop(parts);
+                    drop(parts);
```

Comment:
Is this necessary?. Does it drop it at the end of the statement anyways?

Reply:
I think I just paranoidly put it in there 🙄

Reply:
done